sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentTests.java

```diff
@@ -115,26 +114,69 @@ public CompletableFuture<TestAsyncResult> runAsync(Supplier<CompletableFuture<Ma
             var mockMonitor = (MockMonitor) monitor;
-            return this.runner.runAsyncFuture(callback)
+            var helper = new OutputsCaptureHelper(callback);
+            return this.runner.runAsyncFuture(helper::run)
                     .thenApply(ignore -> new TestAsyncResult(
                             ImmutableList.copyOf(mockMonitor.resources),
                             mockEngine.getErrors().stream()
                                     .map(RunException::new)
-                                    .collect(toImmutableList())
-                    ));
+                                    .collect(toImmutableList()),
+                            helper.getOutputs()));
+        }
+
+        public CompletableFuture<TestAsyncResult> tryTestAsync(Consumer<Context> stackFactory) {
+            return this.runAsync(() -> {
+                var context = new TestContext();
+                stackFactory.accept(context);
+                return CompletableFuture.completedFuture(context.getStackOutputs());
+            });
+        }
+
+        public CompletableFuture<ImmutableList<Resource>> testAsync(Consumer<Context> stackFactory) {
+            return this.tryTestAsync(stackFactory).thenApply(result -> {
+                result.throwOnError();
+                return result.resources;
+            });
         }
 
         public static class TestAsyncResult {
             public final ImmutableList<Resource> resources;
             public final ImmutableList<Exception> exceptions;
+            public final Map<String, Output<?>> stackOutputs;
 
-            public TestAsyncResult(ImmutableList<Resource> resources, ImmutableList<Exception> exceptions) {
+            public TestAsyncResult(ImmutableList<Resource> resources,
+                                   ImmutableList<Exception> exceptions,
+                                   Map<String, Output<?>> stackOutputs) {
                 this.resources = resources;
                 this.exceptions = exceptions;
+                this.stackOutputs = stackOutputs;
+            }
+
+            public <T> Output<T> getStackOutput(String name) {
+                if (!this.stackOutputs.containsKey(name)) {
+                    return Output.of(CompletableFuture.failedFuture(new StackOutputNotFoundException(name)));
+                }
+                return (Output<T>) this.stackOutputs.get(name);
+            }
+
+            public void throwOnError() {
+                if (!this.exceptions.isEmpty()) {
+                    throw new RunException(String.format("Error count: %d, errors: %s",
+                            this.exceptions.size(), this.exceptions.stream()
+                                    .map(Throwable::getMessage)
+                                    .collect(Collectors.joining(", "))
+                    ));
+                }
             }
         }
     }
 
+    private static final class StackOutputNotFoundException extends RuntimeException {
```

Comment:
This one is a bit weird, do we have any situation in main code (non-test) where we're trying to get an output that's not there? What do we throw there? Perhaps when querying a StackReference?

Reply:
looks like it is unused currently

Reply:
I don't think so. We probably should throw in the StackReference, currently we return `null` from what I can see :|